File: Source/Aladdin.ASN1/ASN1/ChoiceCreator.cs
using System;
using System.Reflection;

namespace Aladdin.ASN1
{
	///////////////////////////////////////////////////////////////////////////
	// Фабрика создания фабрик альтернатив
	///////////////////////////////////////////////////////////////////////////
	public class ChoiceCreator<T> where T : IObjectFactory, new()
	{
		// экземпляр фабрики по умолчанию
		public IObjectFactory Factory() { return new T(); }

		// экземпляр фабрики
		public T Factory(params object[] args) 
		{
			// выделить память для типов аргументов
			Type[] types = new Type[args.Length];  

			// указать типы аргументов
			for (int i = 0; i < args.Length; i++) types[i] = args[i].GetType();
 
			// найти соответствующий конструктор
			ConstructorInfo constructor = typeof(T).GetConstructor(types);  

			// вызвать конструктор
			try { return (T)constructor.Invoke(args); }

            // обработать исключение конструктора
            catch (TargetInvocationException e) { throw e.InnerException; }
		}
	}
}

File: Source/Aladdin.ASN1.Java/src/aladdin/asn1/ChoiceCreator.java
package aladdin.asn1;
import java.lang.reflect.*; 

///////////////////////////////////////////////////////////////////////////
// Фабрика создания фабрик альтернатив
///////////////////////////////////////////////////////////////////////////
public final class ChoiceCreator
{
    private final Class<? extends Choice> type; // тип объекта
    
    // конструктор
    public ChoiceCreator(Class<? extends Choice> type) { this.type = type; }
    
    // экземпляр фабрики
    public final Choice factory(Object... args)
    {
        // выделить память для типов аргументов
		Class<?>[] types = new Class<?>[args.length];

		// указать типы аргументов
		for (int i = 0; i < args.length; i++) types[i] = args[i].getClass();
		try { 
			// найти соответствующий конструктор
			Constructor<?> constructor = type.getConstructor(types);

			// вызвать конструктор
			return (Choice)constructor.newInstance(args); 
		}
        // при ошибке выбросить исключение
        catch (Exception e) { throw new RuntimeException(e); }
    }
}